core/src/main/java/kafka/server/share/SharePartition.java

```diff
@@ -1525,6 +1537,24 @@ private Optional<Throwable> acknowledgeCompleteBatch(
         return Optional.empty();
     }
 
+    protected void updateLatestFetchOffsetMetadata(LogOffsetMetadata fetchOffsetMetadata) {
+        lock.writeLock().lock();
+        try {
+            latestFetchOffsetMetadata = fetchOffsetMetadata;
```

Comment:
We need to reset `latestFetchOffsetMetadata` every time `endOffset` changes.

Reply:
IIUC, everytime we fetch new records from `readFromLog`, I utilize the `LogOffsetMetadata` object from the response to update the `latestFetchOffsetMetadata` object while doing the acquire. Hence, this way I have the most recent fetched offset's metadata that I always update through `acquire` method. Most of the time, it is going to be the endOffset's metadata itself, as soon as new data is produced to the topic partition. Anything else that I need to do here?

Reply:
Hmm, if a previously acquired batch times out, SharePartition will change endOffset. This means that the cached latestFetchOffsetMetadata no longer matches the next fetch offset, right?

Reply:
mkaes sense, IIUC, whenever there is an acknowledge or acquisition lock timeout or release records on session close, I should update `latestFetchOffsetMetadata` in `SharePartition` to `Optional.empty()`.

Reply:
Hi @junrao, understood, so if there is any calls to acknowledge/acquisition lock timeout/release acquired records on session close, I should update the latestFetchOffsetMetadata to `Optional.empty()`. Then when the next tryComplete call comes, it will update the `latestFetchOffsetMetadata` and we will have the most recent result. Or is my understanding incorrect?

Reply:
I was thinking about an alternative approach by maintaining `latestFetchOffsetMetadata` every time we update `endOffset`. In the common case, we move `endOffset` to `lastOffset` of an acked batch. The file position in `latestFetchOffsetMetadata` can just be updated by adding the batch size to the file position of the previous `latestFetchOffsetMetadata`. This way, in the common case, we only need to call `readFromLog` once per fetch request, instead of twice (once for getting `latestFetchOffsetMetadata` and another for getting the data) in the current approach.

Reply:
Hi @junrao , i am not sure if I understand the approach completely. IIUC, 
1. I agree with the common case where during `onComplete` we can update the file position by directly adding the batch size. But, there are cases where we return true from `isMinBytesSatisfied` if `fetchOffsetMetadata` was on a different segment than the `endOffsetMetadata` ([code reference](https://github.com/apache/kafka/pull/17539/files#diff-d835cdc01e77905316584ce9e6e21a060cb3d36efa717d4b822b16744e4d713aR296)). In those case, we don't know the number of bytes that got accumulated in the response. Same goes for the case `fetchOffsetMetadata.messageOffset > endOffsetMetadata.messageOffset`, how do we know whta is the number of bytes to add?
2. I am assuming that with this approach, we do not need to do any handling during acquisition lock timeout/acknowledgements/release acquired records on session close. Please correct me if I am wrong.
PS - Thanks a lot for taking out the time to explain me and reviewing this PR.

Reply:
@adixitconfluent : Here is the rough idea.
1. If endOffset advances forward, we incrementally update its file position by the size of batches going forwarded.
2. The tricky thing is how the offset metadata picks up a new segment being rolled. As we increase the file position, endOffset will eventually reach the baseOffset of the next segment. This means that the next fetch request will be satisfied immediately since the HWM is on a different segment. When we acquire the data (for batches at the beginning of the segment), we can check if the offset metadata in the fetch data has the same offset as endOffset but on a different segment. If so, we update the offset segment of endOffset.
3. If the endOffset goes backward (due to timeout/acknowledgements/release) or endOffset is being initialized for the first time, we just call readFromLog to get the offset metadata.
4. `tryComplete` will have the same logic to deal with the uncommon cases where the offset metadata is not available or the offset metadata is on the same segment. The only difference is that it won't update latestFetchOffsetMetadata any more since the update happens when endOffset changes.

Reply:
While this alternative approach is more efficient, it's probably also more complicated. So, it's also ok to just take the current approach to start with. In the current approach, (1) if any call moves `endOffset`, we reset the `latestFetchOffsetMetadata` to Optional.empty(). In `tryComplete`, if `latestFetchOffsetMetadata` is empty, we call `readFromLog` and update `latestFetchOffsetMetadata`.

Reply:
Hi @junrao , agreed with the simpler approach. I have made the following changes in my latest commit- 
> In the current approach, (1) if any call moves endOffset, we reset the latestFetchOffsetMetadata to Optional.empty()

I reset the `latestFetchOffsetMetadata` to Optional.empty() if -
1. `acquire()` results in non-empty acquired records in `ShareFetchUtils`.
2. acquisition lock timeout is called.
3. release acquired records on session close is called.
I haven't made the change in `acknowledge()` method of `SharePartition`, since in the common case all the `ACQUIRED` records will moved to `ACKNOWLEDGED` state and endOffset doesn't change then.

> In tryComplete, if latestFetchOffsetMetadata is empty, we call readFromLog and update latestFetchOffsetMetadata

This functionality has been added in previous commits.

Please review my PR whenever you can. Thanks!